### DISCUSSION TITLE
What do I have to take care of when updating to version 7?

### DISCUSSION THREAD

**Discussion by Tchiller** [deployphp/deployer#2271] [Help needed]
We have updated to version 7 today but we have encountered problems. 

What we managed so far was to replace the abandoned `deployer/recipes` package by calls to e.g. `contrib/cachetools.php`
Then we have changed all occurrences from `hostname` to `alias` in our `deploy.php`. This would look like

```
// Hosts
host('beta')
	->alias('server.domain.com')
	->user('deploy')
	->set('deploy_path', '/var/www/beta')
	->set('cachetool', '/var/run/php7-fpm-beta.sock');

host('production')
	->alias('server.domain.com')
	->user('deploy')
	->set('deploy_path', '/var/www/production')
	->set('cachetool', '/var/run/php7-fpm-production.sock');
```

and we obtain a "successful" deploy (speaking of CI logging), which does not actually deploy but yields a warning ` Call to undefined method Deployer\Host\Host::alias()`

Maybe the Upgrade.md is not clear enough or I am just not getting something. Thank you.

**Comment by antonmedv** [comment]
UPGRADE.md docs soon will be done. For now, here is what correct definition:
```php
host('beta')
	->setHostname('server.domain.com')
	->setRemoteUser('deploy')
	->set('deploy_path', '/var/www/beta')
	->set('cachetool', '/var/run/php7-fpm-beta.sock');
```

So migration is this:
|v6 | v7|
|---|---|
|`hostname`|`alias`|
|`real_hostname`|`hostname`|

**Reply by Tchiller** [reply]
Thank you very much, @antonmedv. We got quite close with the changes shown by @chriswalg but now, even in beta 4, we face the following error: 

`Argument 1 passed to Deployer\Task\ScriptManager::doGetTasks() must be of the type string, null given, called in ./vendor/deployer/deployer/src/Task/ScriptManager.php on line 70`

As the strict parameters are new in v 7, this might be some problem of the `ScriptManager`, `TaskCollection` or `GroupTask` class. Please let me know, if I should file a bug or if you can see a problem in our `deploy.php` shown below. Thanks.

**Reply by antonmedv** [reply]
Please show you full deploy.php file.

**Reply by chriswalg** [reply]
https://gitlab.com/foodsharing-dev/foodsharing/-/blob/master/deploy.php

**Reply by chriswalg** [reply]
For our composer files:
https://gitlab.com/foodsharing-dev/foodsharing/-/tree/master/deployer

**Reply by antonmedv** [reply]
![deploy php · master · foodsharing-dev  foodsharing · GitLab](https://user-images.githubusercontent.com/141232/98392968-f126e100-2058-11eb-851e-e04a3db93048.png)
This us bot right. Error tells exactly what.

**Reply by antonmedv** [reply]
Correct:
```php
<?php
namespace Deployer;

require 'recipe/common.php';
require 'contrib/cachetool.php';

// Project name
set('application', 'foodsharing');

// Project repository
set('repository', 'git@gitlab.com:foodsharing-dev/foodsharing.git');

// Shared files/dirs between deploys
set('shared_files', ['config.inc.prod.php']);
set('shared_dirs', ['images', 'data', 'tmp']);

// Writable dirs by web server
set('writable_dirs', ['tmp', 'cache', 'var']);
set('http_user', 'www-data');
set('remote_user', 'deploy');
set('deploy_path', '/var/www/{{alias}}');
set('cachetool', '/var/run/php7-fpm-{{alias}}.sock');

// default timeout of 300 was failing sometimes
set('default_timeout', 600);

// Hosts
host('beta')
	->setHostname('dragonfruit.foodsharing.network');

host('production')
	->setHostname('dragonfruit.foodsharing.network');

// Tasks
desc('Create the revision information');
task('deploy:create_revision', function () {
	$revision = input()->getOption('revision');
	cd('{{release_path}}');
	run("./scripts/generate-revision.sh $revision");
});

task('deploy:update_code', function () {
	upload(__DIR__ . '/', '{{release_path}}', [
		'--exclude=.git',
		'--exclude=client',
		'--exclude=migrations',
		'--exclude=deployer',
		'--compress-level=9'
	]);
});

task('deploy:cache:warmup', function () {
	run('FS_ENV=prod {{release_path}}/bin/console cache:warmup -e prod');
})->desc('Warmup symfony cache');

desc('Deploy your project');
task('deploy', [
	'deploy:info',
	'deploy:setup',
	'deploy:lock',
	'deploy:release',
	'deploy:update_code',
	'deploy:writable',
	'deploy:shared',
	'deploy:clear_paths',
	'deploy:create_revision',
	'deploy:cache:warmup',
	'deploy:symlink',
        'cachetool:clear:opcache',
	'deploy:unlock',
	'deploy:cleanup',
	'deploy:success'
]);

// [Optional] If deploy fails automatically unlock.
after('deploy:failed', 'deploy:unlock');
```

**Reply by chriswalg** [reply]
@antonmedv Thank you. Your deploy.php is online. Deploy works, but we get an error in cachetool:

`task deploy:setup
task deploy:lock
task deploy:release
task deploy:update_code
[beta] 25567/26603 [≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡>-]  96% 19 secs/20 secs
[beta] 25770/26603 [≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡>]  96% 20 secs/21 secs
[beta] 25830/26603 [≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡>]  97% 21 secs/22 secs
[beta] 25981/26603 [≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡>]  97% 22 secs/23 secs
[beta] 26603/26603 [≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡≡] 100% 22 secs/22 secs
task deploy:writable
task deploy:shared
task deploy:clear_paths
task deploy:create_revision
task deploy:cache:warmup
task deploy:symlink
task cachetool:clear:opcache
[beta]  error  in cachetool.php on line 57:
[beta] run command -v 'cachetool.phar' || which 'cachetool.phar' || type -p 'cachetool.phar'
[beta] exit code 1 (General error)
task deploy:failed
task deploy:unlock
Cleaning up file based variables
00:01
ERROR: Job failed: exit code 1`

https://gitlab.com/foodsharing-dev/foodsharing/-/jobs/835344131

**Reply by antonmedv** [reply]
Download phar.

**Reply by chriswalg** [reply]
@antonmedv Should I add it in deployer/composer.json ?

```
{
    "require-dev": {
        "deployer/deployer": "~7.0.0-beta.4",
        "gordalina/cachetool": "~5.1.3"
    }
}
```

**Reply by chriswalg** [reply]
I tought it is downloaded in https://github.com/deployphp/deployer/blob/master/contrib/cachetool.php#L60

**Reply by antonmedv** [reply]
I maybe a bug.

**Comment by chriswalg** [comment]
Thanks! This changes works for us:

```
Index: deploy.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- deploy.php	(revision 1919706eb47de7574f05ed02da01df4919d1f3c6)
+++ deploy.php	(revision dcc674542012ebc6a1bc32915c05843e78e6d55c)
@@ -28,14 +28,14 @@
 
 // Hosts
 host('beta')
-	->alias('server.domain.com')
-	->user('deploy')
+	->setHostname('server.domain.com')
+	->setRemoteUser('deploy')
 	->set('deploy_path', '/var/www/beta')
 	->set('cachetool', '/var/run/php7-fpm-beta.sock');
 
 host('production')
-	->alias('server.domain.com')
-	->user('deploy')
+	->setHostname('server.domain.com')
+	->setRemoteUser('deploy')
 	->set('deploy_path', '/var/www/production')
 	->set('cachetool', '/var/run/php7-fpm-production.sock');
 
@@ -64,7 +64,7 @@
 desc('Deploy your project');
 task('deploy', [
 	'deploy:info',
-	'deploy:prepare',
+	'deploy:setup',
 	'deploy:lock',
 	'deploy:release',
 	'deploy:update_code',
@@ -73,11 +73,10 @@
 	'deploy:clear_paths',
 	'deploy:create_revision',
 	'deploy:cache:warmup',
-	'deploy:symlink',
-	'cachetool:clear:opcache',
+	after('deploy:symlink', 'cachetool:clear:opcache'),
 	'deploy:unlock',
-	'cleanup',
-	'success'
+	'deploy:cleanup',
+	'deploy:success'
 ]);
 
 // [Optional] If deploy fails automatically unlock.
```
